### ISSUE TITLE
UDPMux causes massive packet loss

### ISSUE THREAD

**Issue by jech** [pion/webrtc#1915]
Testing UDPMux in Galene, I'm seeing absolutely massive packet loss on a local network, on the order of 50-70%.

The code is here: https://github.com/jech/galene/commit/b80e515eb04a8326336524ea80ecf711a3013293

**Comment by OrlandoCo** [comment]
@jech Have you tried to increase OS UDP buffer??

**Comment by jech** [comment]
No.  My current hypothesis is that it's the same issue as https://github.com/pion/webrtc/issues/1356, which is apparently due to having multiple local addresses on a single local socket; that's going to happen on double-stack hosts, as well as on multihomed hosts.

**Comment by davidzhao** [comment]
@jech we are using this (and TCPMux) with LiveKit and only see packet loss when UDP buffer gets overwhelmed. (increasing it gets rid of the loss for us).

https://github.com/livekit/livekit-server/blob/master/pkg/rtc/config.go#L70

Do you want to see if you can repro [with LiveKit](https://docs.livekit.io/guides/getting-started/#start-the-server)? I'm wondering there's something unique to your machine's networking stack.

You can start it with docker, and using UDPMux

```
docker run --rm \
  -p 7880:7880 \
  -p 7881:7881 \
  -p 7882:7882/udp \
  -e LIVEKIT_KEYS="<key>: <secret>" \
  livekit/livekit-server \
  --dev \
  --node-ip=<machine-ip>
```

**Comment by jech** [comment]
Are your machines double-stack?

**Comment by davidzhao** [comment]
what is considered double-stack? having both ipv4/6?

**Comment by jech** [comment]
> what is considered double-stack? having both ipv4/6?

Yes.

**Comment by davidzhao** [comment]
With livekit we are using UDP4 with the mux and that could explain the difference. The challenge with dual-stack is ensuring what's advertised to match the dest addr that we sent to. I remember seeing some oddities along the lines of:

- pion sends packet to client <udp4>:clientport
- client sends packet to pion, which got interpreted as <udp6>:clientport
- mux ignores udp6 address since it's different from the address that it sent to.

**Comment by jech** [comment]
I don't know if the issue is the same as https://github.com/pion/webrtc/issues/1356 (which has higher priority for me), but that issue goes away when I disable IPv6 (see https://github.com/pion/webrtc/issues/1356#issuecomment-894376345).  Disabling IPv6 is of course not an option (IPv6 is great for WebRTC, IPv6 gives you a peer-reflexive candidate straight away, without the need to contact a STUN, which noticably reduces the connexion establishment delay).

**Comment by davidzhao** [comment]
> Disabling IPv6 is of course not an option

I agree having IPv6 is nice, but I would question if it's a must have. Is the slight connection speed improvement worth not having ICE/TCP? that is the decision today.

ofc it'd be ideal to fix the underlying issue.

**Comment by jech** [comment]
The workaround is not a simple matter of disabling IPv6 for TCP-ICE — it requieres disabling IPv6 globally on the host.  This means that you'll run into trouble as soon as somebody runs your code on a modern server.

What's more, the issue indicates that the code is buggy.  Until the bug is understood and properly fixed, there's no saying when the code will bite you.  Most probably during an important demo ;-)

**Comment by davidzhao** [comment]
It'll only cause an issue on servers that don't support IPv4. we have not gotten any feedback about this. AFAIK, all major cloud vendors run their machines with dual stack.

But I digress, let's just fix the underlying issue.

**Comment by stv0g** [comment]
Lets move this issue to pion/ice..
